### ISSUE TITLE
dialect detection module

### ISSUE THREAD

**Issue by runammm** [runammm/HabitLink#10]
**Labels:** `feature`, `ai-model`, `audio-processing`, `machine-learning`

### **Issue: Implement AI-based Dialect Detection Module**

This issue outlines the plan to build and integrate an AI model for detecting Korean dialects from speech audio. The implementation is divided into two major phases: **A) Model Fine-Tuning** and **B) Inference Implementation**.

---

### A. Model Fine-Tuning Workflow

This phase covers the steps required to train a custom model for dialect classification. This will be handled in separate scripts outside the main `src` application directory.

#### 1. Overview
As specified in the project proposal and README, we will use a machine learning model to classify dialects based on accent and intonation. The chosen approach is to **fine-tune a pre-trained `Wav2Vec2` model**, which is a state-of-the-art architecture for speech-related tasks.

#### 2. Data Preparation
The user will provide a dataset of labeled Korean dialect audio. This data needs to be pre-processed for training.

* **Input**: A directory structure where each sub-directory is a dialect label and contains the corresponding `.wav` or `.mp3` audio files.
    ```
    /data/dialect_audio/
    ├── gyeongsang/
    │   ├── audio_001.wav
    │   └── audio_002.wav
    ├── jeolla/
    │   └── ...
    └── chungcheong/
        └── ...
    ```
* **Preprocessing Steps**:
    1.  **Audio Conversion**: Ensure all audio files are in a consistent format (e.g., WAV).
    2.  **Resampling**: All audio must be resampled to **16kHz**, as this is the standard rate for most pre-trained speech models.
    3.  **Dataset Creation**: Use the Hugging Face `datasets` library to load the audio folders into a structured `AudioFolder` dataset.
    4.  **Splitting**: Split the dataset into `train`, `validation`, and `test` sets for proper model training and evaluation.

#### 3. Fine-Tuning Process
We will use the Hugging Face `transformers` library to fine-tune the model.

* **Base Model**: We will start from a `Wav2Vec2` model pre-trained on a large Korean corpus for best results (e.g., `kresnik/wav2vec2-large-xlsr-korean` from the Hugging Face Hub).
* **Training Script**: A dedicated Python script (e.g., `scripts/run_dialect_finetuning.py`) will be created to handle:
    * Loading the pre-processed dataset.
    * Loading the pre-trained `Wav2Vec2` model and feature extractor.
    * Mapping the dialect labels to integer IDs.
    * Defining training arguments (`TrainingArguments`).
    * Using the `Trainer` API to run the fine-tuning process.
* **Evaluation**: The model's performance will be measured using the `accuracy` metric on the validation set during training and on the test set after completion.

---

### B. Inference Implementation

This phase covers integrating the fine-tuned model into the main application.

#### 4. Module Design (`src/dialect_analyzer.py`)
A new module will be created to load the trained model and perform predictions.

```python
# src/dialect_analyzer.py

from transformers import pipeline
from typing import Dict

class DialectAnalyzer:
    """
    Uses a fine-tuned speech classification model to detect Korean dialects from audio.
    """
    def __init__(self, model_path: str):
        """
        Initializes the DialectAnalyzer by loading the fine-tuned model.

        Args:
            model_path (str): The local path or Hugging Face Hub ID of the fine-tuned
                              dialect classification model.
        """
        try:
            # The 'audio-classification' pipeline handles all preprocessing automatically.
            self.classifier = pipeline(
                "audio-classification",
                model=model_path
            )
        except Exception as e:
            print(f"Error loading dialect model from {model_path}: {e}")
            self.classifier = None

    def analyze(self, audio_path: str) -> Dict[str, float]:
        """
        Analyzes a single audio file and returns the predicted dialect probabilities.

        Args:
            audio_path (str): The path to the audio file to be analyzed.

        Returns:
            Dict[str, float]: A dictionary of dialect labels and their corresponding
                              confidence scores.
                              Example: {'Gyeongsang': 0.85, 'Seoul': 0.10, 'Jeolla': 0.05}
        """
        if not self.classifier:
            return {"error": "Model not loaded."}
        
        try:
            predictions = self.classifier(audio_path, top_k=5) # Get top 5 predictions
            
            # Format the output into a simple dictionary
            probabilities = {p['label']: round(p['score'], 3) for p in predictions}
            return probabilities
        except Exception as e:
            print(f"Error during dialect analysis: {e}")
            return {"error": str(e)}
```

#### 5. Integration Plan
The `DialectAnalyzer` operates on audio files, similar to the `SpeakerDiarizer`. It will be integrated into the main pipeline to process the recorded audio in parallel with other modules.

**`src/audio_interface.py` (Proposed Changes):**
```python
# src/audio_interface.py (Conceptual change)

# ... (previous imports)
from .dialect_analyzer import DialectAnalyzer # <-- Import the new class
from typing import List, Dict, Any

class AudioInterface:
    def __init__(self, audio_engine: AudioEngine, diarizer: SpeakerDiarizer, 
                 # ... other analyzers
                 dialect_analyzer: DialectAnalyzer): # <-- Add to init
        # ... (assignments for other analyzers)
        self.dialect_analyzer = dialect_analyzer # <-- Add the analyzer instance

    def record_and_process(self, duration: float, custom_keywords: List[str]) -> Dict[str, Any]:
        """
        Records audio and runs it through the full analysis pipeline.
        """
        # Step 1: Record Audio
        input_path = self.audio_engine.record(duration)
        
        # Step 2: Run parallel analyses
        # Text-based analyses
        diarized_transcript = self.diarizer.process(input_path)
        # ... (calls to word_analyzer, speech_rate_analyzer, grammar_analyzer, context_analyzer)

        # Audio-based analyses
        dialect_prediction = self.dialect_analyzer.analyze(input_path) # <-- Call the new analyzer

        # Step 3: Consolidate all results
        return {
            "full_transcript": diarized_transcript,
            # ... (other text-based analysis results)
            "dialect_analysis": dialect_prediction # <-- Add results to output
        }
```

#### 6. Action Items

**[Model Training]**
- [ ] Set up a Python environment with ML libraries (`transformers`, `datasets`, `torch`, `accelerate`).
- [ ] Write a data preprocessing script to prepare the user-provided audio data.
- [ ] Implement the fine-tuning script (`scripts/run_dialect_finetuning.py`).
- [ ] Run the training process and save the best-performing model checkpoint.

**[Application Integration]**
- [ ] Implement the `DialectAnalyzer` class in a new file `src/dialect_analyzer.py`.
- [ ] Integrate the `DialectAnalyzer` into the main application logic (`src/audio_interface.py`).
- [ ] Update the application's configuration to specify the path to the fine-tuned model.
- [ ] Write a test script to verify that the `analyze` method correctly processes an audio file and returns a probability distribution.
